sklearn/preprocessing/_discretization.py

```diff
@@ -267,7 +278,28 @@ def fit(self, X, y=None):
                 bin_edges[jj] = np.linspace(col_min, col_max, n_bins[jj] + 1)
 
             elif self.strategy == "quantile":
-                quantiles = np.linspace(0, 100, n_bins[jj] + 1)
+                # if sample_weight (weight) parameter is defined, then the size
+                # of each bin should be proportional to each weight value (in w)
+                if sample_weight is not None:
+                    weights = (
+                        sample_weight[jj]
+                        if isinstance(sample_weight[0], collections.Sequence)
+                        else sample_weight
+                    )
+                    if sum(weights) != 1:
+                        raise ValueError(
+                            "Sum of weights in fit method should "
+                            "be equal to 1 (current value : {})".format(sum(weights))
+                        )
+                    quantiles = [0]
+                    wt = 0
+                    for wi in weights:
+                        wt += wi
+                        quantiles.append(100 * wt)
+                # if sample_weight parameter is None then each bin should have
+                # the same number of samples
+                else:
+                    quantiles = np.linspace(0, 100, n_bins[jj] + 1)
```

Comment:
Here the right implementation is to use 

`sklearn.utils.stats._weighted_percentile` instead of the `np.percentile` by providing the weight.

Reply:
Due to a misunderstanding of the task from me, I reviewed my code according to your suggestions.

https://github.com/scikit-learn/scikit-learn/pull/22048/commits/85fe2b74ee5fe571b92f113f9dd8182339a4c4aa